src/include/miopen/cartesianprod/problem_description.hpp

```diff
@@ -0,0 +1,297 @@
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+
+#pragma once
+
+#include <miopen/problem_description_base.hpp>
+#include <miopen/activ.hpp>
+#include <miopen/tensor.hpp>
+
+namespace miopen {
+
+struct NetworkConfig;
+
+namespace cartesianprod {
+
+struct FwdProblemDescription : ProblemDescriptionBase
+{
+    FwdProblemDescription(size_t inputCount_,
+                          const TensorDescriptor* const* inputDescs_,
+                          const TensorDescriptor& outputDesc_)
+        : inputDescs(inputDescs_), outputDesc(outputDesc_), inputCount(inputCount_)
+    {
+        IsValidNumInputs();
+        IsSameType();
+        IsValidDims();
+        IsAllPacked();
+    }
+
+    const TensorDescriptor& GetInputDesc(size_t i) const
+    {
+        if(i >= inputCount)
+        {
+            MIOPEN_THROW(miopenStatusBadParm, "CartesianProdForward: Invalid tensor index.");
+        }
+        return *inputDescs[i];
+    }
+    const TensorDescriptor& GetOutputDesc() const { return outputDesc; }
+    size_t GetInputCount() const { return inputCount; }
+
+    bool IsValidNumInputs() const
+    {
+        if(inputCount < 1)
+        {
+            MIOPEN_THROW(miopenStatusBadParm,
+                         "CartesianProdForward: At least one input tensors are required.");
+        }
+        return true;
+    }
+
+    bool IsSameType() const
+    {
+        const auto dtype = outputDesc.GetType();
+        for(int i = 0; i < inputCount; i++)
+        {
+            if(inputDescs[i]->GetType() != dtype)
+            {
+                MIOPEN_THROW(miopenStatusBadParm,
+                             "CartesianProdForward: Tensor types do not match.");
+            }
+        }
+        return true;
+    }
+
+    bool IsValidDims() const
+    {
+        for(int i = 0; i < inputCount; i++)
+        {
+            if(inputDescs[i]->GetLengths().size() != 1)
+            {
+                MIOPEN_THROW(miopenStatusBadParm, "CartesianProdForward: Invalid input dim.");
+            }
+        }
+        if(inputCount == 1)
+        {
+            if(outputDesc.GetLengths().size() != 1)
+            {
+                MIOPEN_THROW(miopenStatusBadParm, "CartesianProdForward: Invalid output dim.");
+            }
+        }
+        else
+        {
+            if(outputDesc.GetLengths().size() != 2)
+            {
+                MIOPEN_THROW(miopenStatusBadParm, "CartesianProdForward: Invalid output dim.");
+            }
+            size_t outputDim0 = 1;
+            for(int i = 0; i < inputCount; i++)
+            {
+                outputDim0 *= inputDescs[i]->GetLengths()[0];
+            }
+            if(outputDesc.GetLengths()[0] != outputDim0 || outputDesc.GetLengths()[1] != inputCount)
+            {
+                MIOPEN_THROW(miopenStatusBadParm, "CartesianProdForward: Invalid output dim.");
+            }
+        }
+
+        return true;
+    }
+
+    bool IsAllContiguous() const
```

Comment:
You may don't need this function anymore.

Reply:
resolved